modules/auxiliary/cloud/aws/enum_ssm.rb

```diff
@@ -0,0 +1,153 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'aws-sdk-ssm'
+
+class MetasploitModule < Msf::Auxiliary
+  include Rex::Proto::Http::WebSocket::AmazonSsm
+  include Msf::Auxiliary::CommandShell
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name'        => 'Amazon Web Services EC2 instance enumeration',
+        'Description' => %q(
+                          Provided AWS credentials, this module will call the authenticated
+                          API of Amazon Web Services to list all SSM-enabled EC2 instances
+                          accessible to the account. Once enumerated as SSM-enabled, the
+                          instances can be controlled using out-of-band WebSocket sessions
+                          provided by the AWS API (nominally, privileged out of the box).
+                          This module provides not only the API enumeration identifying EC2
+                          instances accessible via SSM with given credentials, but enables
+                          session initiation for all identified targets (without requiring
+                          target-level credentials). The module also provides an EC2 ID
+                          filter and a limiting throttle to prevent session stampedes or
+                          expensive messes.
+                         ),
+        'Author'      => [
+          'Aaron Soto <aaron.soto@rapid7.com>', # EC2 enum module
+          'RageLtMan <rageltman[at]sempervictus>' # SSM stuff
+        ],
+        'License'     => MSF_LICENSE
+      )
+    )
+
+    register_options(
+      [
+        OptBool.new('GET_SSM_SESSION', [true, 'Automatically get SSM sessions once found', false]),
+        OptInt.new('LIMIT', [false, 'Only return the specified number of results from each region']),
+        OptString.new('FILTER_EC2_ID', [false, 'Look for specific EC2 instance ID']),
+        OptString.new('REGION', [false, 'AWS Region (eg. "us-west-2")']),
+        OptString.new('ACCESS_KEY_ID', [true, 'AWS Access Key ID (eg. "AKIAXXXXXXXXXXXXXXXX")', '']),
```

Comment:
Do we ever check to make sure these values are sane before using them?  We're defaulting to '', but we should probably make sure that a user enters a value before using it, though I might be missing that check.

Reply:
The datastore enforces the required values (the `true` at the beginning of the array vs the `false` for optional params).
We can add validators for AWS key formats i guess, but that seems like spoonfeeding so didn't occur to me :smile:
We probably want an AWS mixin to handle that anyway... I noticed when i did the Aux module, that my Opt names differ from the existing ones.